110116 accreditation api app code

opticbob opened the pull request:
Keep your PR as a Draft until it's ready for Platform review. A PR is ready for Platform review when it has a teammate approval and tests, linting, and settings checks pass CI. See [these tips](https://depo-platform-documentation.scrollhelp.site/developer-docs/vets-api-pr-tips) on how to avoid common delays in getting your PR merged.

## Summary

- *This work is behind a feature toggle (flipper): NO*
- *(Summarize the changes that have been made to the platform)*: This PR adds all new code to access the GCLAWS Accreditation API and uses that to update the `AccreditedIndividual` agents and attorney types.  It is broken up into three main sets of changes:
  - This adds an Accredited API Entity Count model that will be used to implement data safety measures for our use of that API.  The logic itself gets the current counts from the API and compares it to a record of the previous counts or counts the actual records in the database if necessary.
  - This PR uses the GCLAWS Accreditation API to keep the AccreditedIndividual records up to date. It creates new records, updates existing ones, and deletes old records. The job itself iterates through agents and attorneys, checks to see if there is enough data to proceed and if so makes sure all records are in place with find_or_create and enqueues records with changed addresses for additional address validation. After all of that all existing records that weren't in the daily data set from the API are deleted.
  - This PR contains the code to schedule the Accreditation API update job and to reprocess the job manually if needed. There is a small change to periodic jobs to schedule the daily job and a rake task that can manually enqueue the main processing job to override the data safety measures.  The biggest chunk of code is a sidekiq job that takes accredited individual data in json and uses the Address Validation API to get better addresses along with a lat/long pair we use in Find A Rep. That job has lots of retry logic built in to try to get the (user submitted) addresses to match with a real address within their data set.
- *(If bug, how to reproduce)*
- *(What is the solution, why is this the solution?)*: This lets us move away from our file parsing based current process and removes any manual tasks needed by the team.
- *(Which team do you work for, does your team own the maintenance of this component?)*: FAR/ARM.  We own the maintenance.
- *(If introducing a flipper, what is the success criteria being targeted?)*

## Related issue(s)

- https://github.com/department-of-veterans-affairs/va.gov-team/issues/110116
- *Link to previous change of the code/bug (if applicable)*
- *Link to epic if not included in ticket*

## Testing done

- [x] *New code is covered by unit tests* **In https://github.com/department-of-veterans-affairs/vets-api/pull/22912**
- *Describe what the old behavior was prior to the change*: This is all new code.
- *Describe the steps required to verify your changes are working as expected. Exclusively stating 'Specs run' is NOT acceptable as appropriate testing*: This is difficult to test outside of the staging environment, other environments won't have access to the Accreditation API.  Once it is deployed to Staging, inspect the `AccreditedIndividual `records to ensure they're being written by the job.
- *If this work is behind a flipper:*
  - *Tests need to be written for both the flipper on and flipper off scenarios. [Docs](https://depo-platform-documentation.scrollhelp.site/developer-docs/feature-toggles-guide#Featuretogglesguide-Backendexample).*
  - *What is the testing plan for rolling out the feature?*

## Screenshots / Video
This shows processing agents and orgs in staging.  The code was copied into the staging console.  It was modified to remove any modification of the database and to log out to the console.
https://github.com/user-attachments/assets/18de6784-7e39-4e65-b987-f0b8f79a8b9e
<details><summary>How to manually test</summary>

```ruby
# frozen_string_literal: true

module RepresentationManagement
  # Tracks and validates counts of accredited entities from the GCLAWS API
  #
  # This class stores historical count data for different types of accredited
  # entities (agents, attorneys, representatives, veteran service organizations)
  # and enforces validation rules to prevent dramatic decreases in counts which
  # might indicate data integrity issues.
  #
  # @example Saving new API counts
  #   counter = AccreditationApiEntityCount.new
  #   counter.save_api_counts
  #
  # @example Checking if a count is valid
  #   counter = AccreditationApiEntityCount.new
  #   counter.valid_count?('agents')
  class AccreditationApiEntityCount < ApplicationRecord
    self.table_name = 'accreditation_api_entity_counts'

    # Entity types supported by the GCLAWS API
    TYPES = RepresentationManagement::GCLAWS::Client::ALLOWED_TYPES

    # The maximum allowed percentage decrease between consecutive counts
    # If the new count is less than (previous count * (1 - DECREASE_THRESHOLD)),
    # the count is considered invalid and requires manual review
    DECREASE_THRESHOLD = 0.20 # 20% maximum decrease allowed

    # Retrieves current counts from the API and saves them to the database
    # if they pass validation checks.
    #
    # @return [Boolean] true if save was successful, false otherwise
    def save_api_counts
      TYPES.each do |type|
        send("#{type}=", current_api_counts[type]) if valid_count?(type, notify: false)
      end

      save!
    rescue => e
      log_error("Error saving API counts: #{e.message}")
    end

    # Validates that a new count doesn't decrease too dramatically from the previous count according to the
    # DECREASE_THRESHOLD.
    #
    # @param type [Symbol] The entity type to validate (:agents, :attorneys, etc.)
    # @param notify [Boolean] Whether to send notifications if threshold is exceeded
    # @return [Boolean] true if count is valid, false otherwise
    def valid_count?(type, notify: true)
      previous_count = current_db_counts[type]
      new_count = current_api_counts[type]

      # If no previous count exists, allow the update
      return true if previous_count.nil? || previous_count.zero?

      # If new count is greater or equal, allow the update
      return true if new_count >= previous_count

      # Calculate decrease percentage
      decrease_percentage = (previous_count - new_count).to_f / previous_count

      if decrease_percentage > DECREASE_THRESHOLD
        # Log to Slack and don't update
        notify_threshold_exceeded(type, previous_count, new_count, decrease_percentage, DECREASE_THRESHOLD) if notify
        false
      else
        true
      end
    end

    private

    # @return [RepresentationManagement::GCLAWS::Client] The client for GCLAWS API calls
    def client
      RepresentationManagement::GCLAWS::Client
    end

    # @return [Hash] The current counts from the API for all entity types
    def current_api_counts
      @current_api_counts ||= get_counts_from_api
    end

    # Fetches entity counts from the GCLAWS API
    #
    # @return [Hash] A hash mapping entity types to their counts
    def get_counts_from_api
      counts = {}
      TYPES.each do |type|
        # We're fetching with a page size of 1 to get the fastest possible response for the total count
        counts[type] = client.get_accredited_entities(type:, page: 1, page_size: 1).body['totalRecords']
      rescue => e
        log_error("Error fetching count for #{type}: #{e.message}")
      end
      counts
    end

    # @return [Hash] The most recent counts from the database for all entity types
    def current_db_counts
      @current_db_counts ||= get_counts_from_db
    end

    # Gets the most recent counts from the database, starting with records of this model and falling back to counting
    # records in the database if no previous counts exist
    #
    # @return [Hash] A hash mapping entity types to their counts
    def get_counts_from_db
      latest_counts = RepresentationManagement::AccreditationApiEntityCount.order(created_at: :desc).first
      # All of the counts fallback to counting records in the database if the latest counts are not available.
      {
        agents: latest_counts&.agents || individual_count('claims_agent'),
        attorneys: latest_counts&.attorneys || individual_count('attorney'),
        representatives: latest_counts&.representatives || individual_count('representative'),
        veteran_service_organizations: latest_counts&.veteran_service_organizations || AccreditedOrganization.count
      }
    end

    # Counts the number of accredited individuals of a specific type
    #
    # @param type [String] The individual type ('claims_agent', 'attorney', etc.)
    # @return [Integer] The count of individuals of the specified type
    def individual_count(type)
      AccreditedIndividual.where(individual_type: type).count
    end

    # Notification and logging methods
    # Notifies stakeholders when an entity count decreases beyond the threshold
    #
    # @param rep_type [Symbol] The entity type that exceeded the threshold
    # @param previous_count [Integer] The previous count
    # @param new_count [Integer] The new count
    # @param decrease_percentage [Float] The calculated decrease percentage
    # @param threshold [Float] The threshold that was exceeded
    def notify_threshold_exceeded(rep_type, previous_count, new_count, decrease_percentage, threshold)
      message = "⚠️ AccreditationApiEntityCount Alert: #{rep_type.to_s.humanize} count decreased beyond threshold!\n" \
                "Previous: #{previous_count}\n" \
                "New: #{new_count}\n" \
                "Decrease: #{(decrease_percentage * 100).round(2)}%\n" \
                "Threshold: #{(threshold * 100).round(2)}%\n" \
                'Action: Update skipped, manual review required'

      log_to_slack_threshold_channel(message)
      log_error("AccreditationApiEntityCount threshold exceeded for #{rep_type}, previous: #{previous_count}, " \
                "new: #{new_count}, decrease: #{(decrease_percentage * 100).round(2)}%")
    end

    # Sends a notification to the Slack channel
    #
    # @param message [String] The message to send to Slack
    def log_to_slack_threshold_channel(message)
      slack_client = SlackNotify::Client.new(webhook_url: Settings.claims_api.slack.webhook_url,
                                             channel: '#benefits-representation-management-notifications',
                                             username: 'RepresentationManagement::AccreditationApiEntityCount')
      slack_client.notify(message)
    end

    # Logs an error message to the Rails logger
    #
    # @param message [String] The error message to log
    def log_error(message)
      Rails.logger.error("RepresentationManagement::AccreditationApiEntityCount error: #{message}")
    end
  end
end

require 'sidekiq'

module RepresentationManagement
  # This is the first job in a two job process for updating accredited entities.
  # Processes and updates accredited entities (agents and attorneys) from the GCLAWS API
  #
  # This Sidekiq job fetches data about accredited agents and attorneys from the GCLAWS API,
  # creates or updates records in the database, validates addresses through the VAProfile
  # address validation service, and removes records that are no longer present in the API.
  # That address validation is done in the second job, RepresentationManagement::AccreditedIndividualsUpdate.
  #
  # The job includes data validation to prevent large decreases in entity counts, which might
  # indicate data quality issues. This validation can be bypassed for specific entity types
  # using the force_update_types parameter.
  #
  # @example Enqueue the job to process all entity types
  #   RepresentationManagement::AccreditedEntitiesQueueUpdates.perform_async
  #
  # @example Force update for a specific entity type
  #   RepresentationManagement::AccreditedEntitiesQueueUpdates.perform_async(['agents'])
  #
  # @example Force update for multiple entity types
  #   RepresentationManagement::AccreditedEntitiesQueueUpdates.perform_async(['agents', 'attorneys'])
  class AccreditedEntitiesQueueUpdates
    include Sidekiq::Job

    # Maximum allowed decrease percentage for entity counts before updates are blocked
    DECREASE_THRESHOLD = RepresentationManagement::AccreditationApiEntityCount::DECREASE_THRESHOLD

    # Number of records to process in each address validation batch
    SLICE_SIZE = 30

    # Define a configuration map for entity types
    ENTITY_CONFIG = {
      'agents' => {
        singular_type: 'agent',
        individual_type: 'claims_agent',
        responses_var: :@agent_responses,
        ids_var: :@agent_ids,
        json_var: :@agent_json_for_address_validation,
        validation_description: 'Batching agent address updates from GCLAWS Accreditation API'
      },
      'attorneys' => {
        singular_type: 'attorney',
        individual_type: 'attorney',
        responses_var: :@attorney_responses,
        ids_var: :@attorney_ids,
        json_var: :@attorney_json_for_address_validation,
        validation_description: 'Batching attorney address updates from GCLAWS Accreditation API'
      }
    }.freeze

    # Main job method that processes accredited entities
    #
    # @param force_update_types [Array<String>] Optional array of entity types to force update
    #   regardless of count validation ('agents', 'attorneys')
    # @return [void]
    def perform(force_update_types = [])
      # The force_update_types are the accredited entity types that should be updated regardless of the current counts.
      @force_update_types = force_update_types
      @agent_responses = []
      @attorney_responses = []
      @agent_ids = []
      @attorney_ids = []
      @agent_json_for_address_validation = []
      @attorney_json_for_address_validation = []
      @entity_counts = RepresentationManagement::AccreditationApiEntityCount.new

      # Don't save fresh API counts if updates are forced
      @entity_counts.save_api_counts unless @force_update_types.any?
      process_entity_type('agents')
      process_entity_type('attorneys')
      delete_removed_accredited_individuals
    rescue => e
      log_error("Error in AccreditedEntitiesQueueUpdates: #{e.message}")
    end

    private

    # Processes entities of a specific type based on count validation and force update settings
    #
    # @param entity_type [String] The type of entity to process ('agents' or 'attorneys')
    # @return [void]
    def process_entity_type(entity_type)
      # Don't process if we are forcing updates for other types
      return if @force_update_types.any? && @force_update_types.exclude?(entity_type)

      if @entity_counts.valid_count?(entity_type) || @force_update_types.include?(entity_type)
        if entity_type == 'agents'
          update_agents
          validate_agent_addresses
        else # attorneys
          update_attorneys
          validate_attorney_addresses
        end
      else
        entity_display = entity_type.capitalize
        log_error("#{entity_display} count decreased by more than #{DECREASE_THRESHOLD * 100}% - skipping update")
      end
    end

    # Fetches agent data from the GCLAWS API and updates database records
    #
    # @return [void]
    def update_agents
      update_entities('agents')
    end

    # Fetches attorney data from the GCLAWS API and updates database records
    #
    # @return [void]
    def update_attorneys
      update_entities('attorneys')
    end

    # Generic method to update entities of a specific type
    #
    # @param entity_type [String] The type of entity to update ('agents' or 'attorneys')
    # @return [void]
    def update_entities(entity_type)
      config = ENTITY_CONFIG[entity_type]
      page = 1

      loop do
        response = client.get_accredited_entities(type: entity_type, page:)
        entities = response.body['items']
        break if entities.empty?

        instance_variable_get(config[:responses_var]) << entities
        entities.each { |entity| handle_entity_record(entity, config) }
        page += 1
      end
    end

    # Removes AccreditedIndividual records that are no longer present in the GCLAWS API
    #
    # @return [void]
    def delete_removed_accredited_individuals
      AccreditedIndividual.where.not(id: @agent_ids + @attorney_ids).find_each do |record|
        record.destroy
      rescue => e
        log_error("Error deleting old accredited individual with ID #{record.id}: #{e.message}")
      end
    end

    # Handle an individual entity
    #
    # @param entity [Hash] The entity data from the API
    # @param config [Hash] Configuration for the entity type
    # @return [void]
    def handle_entity_record(entity, config)
      singular_type = config[:singular_type]
      entity_hash = send("data_transform_for_#{singular_type}", entity)

      # Find or create record
      entity_identifier = { individual_type: config[:individual_type], ogc_id: entity['id'] }
      p "entity_identifier: #{entity_identifier}", "entity_hash: #{entity_hash}"
      # record = AccreditedIndividual.find_or_create_by(entity_identifier)

      # # Check if address validation is needed
      # raw_address = send("raw_address_for_#{singular_type}", entity)
      # if record.raw_address != raw_address
      #   json_method = "individual_#{singular_type}_json"
      #   instance_variable_get(config[:json_var]) << send(json_method, record, entity)
      # end

      # # Update record and store ID
      # record.update(entity_hash)
      # instance_variable_get(config[:ids_var]) << record.id
    end

    # Transforms agent data from the GCLAWS API into a format suitable for the AccreditedIndividual model
    #
    # @param agent [Hash] Raw agent data from the GCLAWS API
    # @return [Hash] Transformed data for AccreditedIndividual record
    def data_transform_for_agent(agent)
      data_transform_for_entity(agent, 'claims_agent', {
                                  country_code_iso3: agent['workCountry'],
                                  country_name: agent['workCountry'],
                                  phone: agent['workPhoneNumber'],
                                  email: agent['workEmailAddress'],
                                  raw_address: raw_address_for_agent(agent)
                                })
    end

    # Transforms attorney data from the GCLAWS API into a format suitable for the AccreditedIndividual model
    #
    # @param attorney [Hash] Raw attorney data from the GCLAWS API
    # @return [Hash] Transformed data for AccreditedIndividual record
    def data_transform_for_attorney(attorney)
      data_transform_for_entity(attorney, 'attorney', {
                                  city: attorney['workCity'],
                                  state_code: attorney['workState'],
                                  phone: attorney['workNumber'],
                                  email: attorney['emailAddress']
                                })
    end

    # Base transformation method for both agents and attorneys
    #
    # @param entity [Hash] Raw entity data from the GCLAWS API
    # @param entity_type [String] The type of entity ('claims_agent' or 'attorney')
    # @param extra_attrs [Hash] Additional attributes specific to this entity type
    # @return [Hash] Transformed data for AccreditedIndividual record
    def data_transform_for_entity(entity, entity_type, extra_attrs = {})
      {
        individual_type: entity_type,
        registration_number: entity['number'],
        poa_code: entity['poa'],
        ogc_id: entity['id'],
        first_name: entity['firstName'],
        middle_initial: entity['middleName'].to_s.strip.first,
        last_name: entity['lastName'],
        address_line1: entity['workAddress1'],
        address_line2: entity['workAddress2'],
        address_line3: entity['workAddress3'],
        zip_code: entity['workZip']
      }.merge(extra_attrs)
    end

    # Extracts address data from an entity record with optional extra fields
    #
    # @param entity [Hash] Raw entity data from the GCLAWS API
    # @param extra_fields [Hash] Additional fields to include, with mapping to entity keys
    # @return [Hash] Standardized address data
    def raw_address_from_entity(entity, extra_fields = {})
      {
        address_line1: entity['workAddress1'],
        address_line2: entity['workAddress2'],
        address_line3: entity['workAddress3'],
        zip_code: entity['workZip']
      }.merge(extra_fields.transform_values { |key| entity[key] })
        .transform_keys(&:to_s)
    end

    # Creates a standardized address hash for an agent
    #
    # @param agent [Hash] Raw agent data from the GCLAWS API
    # @return [Hash] Standardized address data
    def raw_address_for_agent(agent)
      raw_address_from_entity(agent, work_country: 'workCountry')
    end

    # Creates a standardized address hash for an attorney
    #
    # @param attorney [Hash] Raw attorney data from the GCLAWS API
    # @return [Hash] Standardized address data
    def raw_address_for_attorney(attorney)
      raw_address_from_entity(attorney, city: 'workCity', state_code: 'workState')
    end

    # Creates a JSON object for an agent's address, used for address validation
    #
    # @param record [AccreditedIndividual] The database record for the agent
    # @param agent [Hash] Raw agent data from the GCLAWS API
    # @return [Hash] JSON structure for address validation
    def individual_agent_json(record, agent)
      individual_entity_json(record, agent, :agent, { city: nil })
    end

    # Creates a JSON object for an attorney's address, used for address validation
    #
    # @param record [AccreditedIndividual] The database record for the attorney
    # @param attorney [Hash] Raw attorney data from the GCLAWS API
    # @return [Hash] JSON structure for address validation
    def individual_attorney_json(record, attorney)
      attorney_raw_address = raw_address_for_attorney(attorney)
      individual_entity_json(
        record,
        attorney,
        :attorney,
        {
          city: attorney_raw_address['city'],
          state: { state_code: attorney_raw_address['state_code'] }
        }
      )
    end

    # Base method to create a JSON object for entity address validation
    #
    # @param record [AccreditedIndividual] The database record for the entity
    # @param entity [Hash] Raw entity data from the GCLAWS API
    # @param entity_type [Symbol] The type of entity (:agent or :attorney)
    # @param additional_fields [Hash] Additional address fields specific to this entity type
    # @return [Hash] JSON structure for address validation
    def individual_entity_json(record, entity, entity_type, additional_fields = {})
      raw_address = send("raw_address_for_#{entity_type}", entity)

      {
        id: record.id,
        address: {
          address_pou: 'RESIDENCE/CHOICE',
          address_line1: raw_address['address_line1'],
          address_line2: raw_address['address_line2'],
          address_line3: raw_address['address_line3'],
          zip_code5: raw_address['zip_code']
        }.merge(additional_fields)
      }
    end

    # Queues address validation jobs for a batch of records
    #
    # @param records_for_validation [Array<Hash>] Records to validate
    # @param description [String] Description for the Sidekiq batch
    # @return [void]
    def validate_addresses(records_for_validation,
                           description = 'Batching address updates from GCLAWS Accreditation API')
      return if records_for_validation.empty?

      delay = 0
      batch = Sidekiq::Batch.new
      batch.description = description

      begin
        batch.jobs do
          records_for_validation.each_slice(SLICE_SIZE) do |individuals|
            json_individuals = individuals.to_json
            RepresentationManagement::AccreditedIndividualsUpdate.perform_in(delay.minutes, json_individuals)
            delay += 1
          end
        end
      rescue => e
        log_error("Error queuing address updates: #{e.message}")
      end
    end

    # Queues address validation jobs for agents
    #
    # @return [void]
    def validate_agent_addresses
      validate_entity_addresses('agents')
    end

    # Queues address validation jobs for attorneys
    #
    # @return [void]
    def validate_attorney_addresses
      validate_entity_addresses('attorneys')
    end

    # Queues address validation jobs for a specific entity type
    #
    # @param entity_type [String] The entity type to validate ('agents' or 'attorneys')
    # @return [void]
    def validate_entity_addresses(entity_type)
      config = ENTITY_CONFIG[entity_type]
      json_var = config[:json_var]
      description = config[:validation_description]

      validate_addresses(
        instance_variable_get(json_var),
        description
      )
    end

    # @return [RepresentationManagement::GCLAWS::Client] The client for GCLAWS API calls
    def client
      RepresentationManagement::GCLAWS::Client
    end

    # Logs an error message to the Rails logger
    #
    # @param message [String] The error message to log
    # @return [void]
    def log_error(message)
      Rails.logger.error("RepresentationManagement::AccreditedEntitiesQueueUpdates error: #{message}")
    end
  end
end

RepresentationManagement::AccreditedEntitiesQueueUpdates.new.perform
```
</details>

## What areas of the site does it impact?
*(Describe what parts of the site are impacted and*if*code touched other areas)*: Mostly the `RepresentationManagement` module.  A small change to `periodic_jobs.rb`

## Acceptance criteria

- [x]  I added unit tests and integration tests for each feature (if applicable). **In https://github.com/department-of-veterans-affairs/vets-api/pull/22912**
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

pmclaren19 commented on modules/representation_management/app/sidekiq/representation_management/accredited_entities_queue_updates.rb:
```diff
@@ -0,0 +1,333 @@
+require 'sidekiq'
+
+module RepresentationManagement
+  # Processes and updates accredited entities (agents and attorneys) from the GCLAWS API
```
can we please update this to also include that this is job #1

pmclaren19 commented on modules/representation_management/app/sidekiq/representation_management/accredited_entities_queue_updates.rb:
```diff
@@ -0,0 +1,333 @@
+  # Processes and updates accredited entities (agents and attorneys) from the GCLAWS API
+  #
+  # This Sidekiq job fetches data about accredited agents and attorneys from the GCLAWS API,
+  # creates or updates records in the database, validates addresses through the VAProfile
```
maybe update this to say it then called job #1 which validates the addresses through the VAProfile

pmclaren19 commented on modules/representation_management/app/sidekiq/representation_management/accredited_individuals_update.rb:
```diff
@@ -0,0 +1,284 @@
+      reps_data = JSON.parse(reps_json)
+      reps_data.each { |rep_data| process_rep_data(rep_data) }
+    rescue => e
+      log_error("Error processing job: #{e.message}", send_to_slack: true)
```
can we log the job id here?

opticbob replied:
I added it in the `log_error `method.

holdenhinkle commented on modules/representation_management/app/sidekiq/representation_management/accredited_entities_queue_updates.rb:
```diff
@@ -0,0 +1,335 @@
+    # Controls the processing of agents based on count validation and force update settings
+    #
+    # @return [void]
+    def process_agents
```
`#process_agents` and `#process_attorneys` are basically the same methods. Can you DRY this up? Create a more generic method that takes the individual type...

opticbob replied:
Yep, done!

holdenhinkle commented on modules/representation_management/app/sidekiq/representation_management/accredited_entities_queue_updates.rb:
```diff
@@ -0,0 +1,335 @@
+    # Fetches agent data from the GCLAWS API and updates database records
+    #
+    # @return [void]
+    def update_agents
```
`#update_agents` and `#update_attorneys` are also basically the same. Can you DRY this up too?

opticbob replied:
Done!

holdenhinkle commented on modules/representation_management/app/sidekiq/representation_management/accredited_entities_queue_updates.rb:
```diff
@@ -0,0 +1,335 @@
+    # @param record [AccreditedIndividual] The database record for the agent
+    # @param agent [Hash] Raw agent data from the GCLAWS API
+    # @return [Hash] JSON structure for address validation
+    def individual_agent_json(record, agent)
```
Same. DRY up `#individual_agent_json` and `#individual_attorney_json` - they're basically the same methods.

opticbob replied:
Done!

holdenhinkle commented on modules/representation_management/app/sidekiq/representation_management/accredited_entities_queue_updates.rb:
```diff
@@ -0,0 +1,335 @@
+    #
+    # @param agent [Hash] Raw agent data from the GCLAWS API
+    # @return [Hash] Transformed data for AccreditedIndividual record
+    def data_transform_for_agent(agent)
```
Same DRY comment for `#data_transform_for_agent` and `#data_transform_for_attorney`.

holdenhinkle commented on modules/representation_management/app/sidekiq/representation_management/accredited_entities_queue_updates.rb:
```diff
@@ -0,0 +1,335 @@
+    # Queues address validation jobs for agents
+    #
+    # @return [void]
+    def validate_agent_addresses
```
Same DRY comment.

opticbob replied:
Done!

pmclaren19 reviewed (dismissed):
LGTM thanks for adding those comments! 

Tested manually with @opticbob in staging console and worked great!

holdenhinkle reviewed (commented):
I made a number of comments about DRYing up the claims agent and attorney version of each method. Apologies if my comments seem out of order - it was like I was review the file from the bottom up.

What about diffing the email and phone?

opticbob commented:
@holdenhinkle I'll look at DRYing up those methods tomorrow morning.  As far as the email and phone, we're not diffing those values.  We're directly updating the db with whatever value the API gives us every day because there is no external validation for those values.

holdenhinkle commented:
>@holdenhinkle I'll look at DRYing up those methods tomorrow morning. As far as the email and phone, we're not diffing those values. We're directly updating the db with whatever value the API gives us every day because there is no external validation for those values.

Ah - you took the same approach as the VSOReloader job - always update. My approach was more surgical - update only when diffs are detected. That's fine - the email and phone updates are on the spot and not in a background job.

opticbob commented on modules/representation_management/app/sidekiq/representation_management/accredited_individuals_update.rb:
```diff
@@ -0,0 +1,286 @@
+    # @param error [Exception] The error string to be logged.
+    def log_error(error, send_to_slack: false)
+      job_id = self.jid
+      message = "RepresentationManagement::AccreditedIndividualsUpdate: #{error} (Job ID: #{job_id})"
```
The `job_id` logging is no longer in place.

opticbob commented:
@holdenhinkle and @pmclaren19 can you take another look?  I've addressed all your feedback and grouped methods by theme.  There are no changes needed to the test PR.

holdenhinkle reviewed (dismissed):
Great job!

opticbob commented on modules/representation_management/app/sidekiq/representation_management/accredited_entities_queue_updates.rb:
```diff
@@ -0,0 +1,370 @@
+    # Removes AccreditedIndividual records that are no longer present in the GCLAWS API
+    #
+    # @return [void]
+    def delete_old_accredited_individuals
```
@holdenhinkle I renamed it, can you re-review?

pmclaren19 commented on modules/representation_management/app/sidekiq/representation_management/accredited_entities_queue_updates.rb:
```diff
@@ -0,0 +1,370 @@
+
+      # Don't save fresh API counts if updates are forced
+      @entity_counts.save_api_counts unless @force_update_types.any?
+      process_entity_type('agents')
```
maybe make these a constant so that you dont have to type out agents and attorneys every time

```
    ENTITY_TYPE = {
      AGENTS: 'AGENTS',
      ATTORNEYS: 'ATTORNEYS',
    }.freeze
```

pmclaren19 replied:
it looks like you have multiple places that are using 'agents' so having an enum would make the code more readable and easier to maintain.

opticbob replied:
Thanks for talking through this with me, I'll make this happen Monday!

pmclaren19 commented on modules/representation_management/app/sidekiq/representation_management/accredited_entities_queue_updates.rb:
```diff
@@ -0,0 +1,370 @@
+    # Fetches agent data from the GCLAWS API and updates database records
+    #
+    # @return [void]
+    def update_agents
```
I would remove this code and in the if statement above in  `process_entity_type ` just do 
`update_entities('agents')`.

opticbob replied:
See above: https://github.com/department-of-veterans-affairs/vets-api/pull/22911/files#r2180546693

pmclaren19 commented on modules/representation_management/app/sidekiq/representation_management/accredited_entities_queue_updates.rb:
```diff
@@ -0,0 +1,370 @@
+      return if @force_update_types.any? && @force_update_types.exclude?(entity_type)
+
+      if @entity_counts.valid_count?(entity_type) || @force_update_types.include?(entity_type)
+        if entity_type == 'agents'
```
I would change this to be 

```
if entity_type == ENTITY_TYPE.AGENTS
  update_entities(ENTITY_TYPE.AGENTS)
  validate_agent_addresses
else
  update_entities(ENTITY_TYPE. ATTORNEYS)
  validate_attorney_addresses
end
```

opticbob replied:
Right now I'm using the separate `update_agents` and `update_attorneys` methods because those methods are set up to be individually tested in the test PR.  I can make this change but I'm not sure how much benefit it has overall.

pmclaren19 replied:
my thinking was that it was less code, easier to read and there wasnt a need for two new methods. If you think this is necessary for testing im open to discuss!

opticbob replied:
Thanks for talking through this with me, I'll make this happen Monday!

pmclaren19 commented on modules/representation_management/app/sidekiq/representation_management/accredited_entities_queue_updates.rb:
```diff
@@ -0,0 +1,370 @@
+    # @param agent [Hash] Raw agent data from the GCLAWS API
+    # @return [Hash] Transformed data for AccreditedIndividual record
+    def data_transform_for_agent(agent)
+      data_transform_for_entity(agent, 'claims_agent', {
```
should this be `agents` or `claims_agent` ?

opticbob replied:
`claims_agent`, it'll be assigned to the `individual_type` of the `AccreditedIndividual`.

pmclaren19 commented on modules/representation_management/app/sidekiq/representation_management/accredited_entities_queue_updates.rb:
```diff
@@ -0,0 +1,370 @@
+    # @param config [Hash] Configuration for the entity type
+    # @return [void]
+    def handle_entity_record(entity, config)
+      singular_type = config[:singular_type]
```
can we just call this entity type since that what you call i everywhere else?

opticbob commented on modules/representation_management/app/sidekiq/representation_management/accredited_entities_queue_updates.rb:
```diff
@@ -0,0 +1,370 @@
+    #
+    # @param agent [Hash] Raw agent data from the GCLAWS API
+    # @return [Hash] Transformed data for AccreditedIndividual record
+    def data_transform_for_agent(agent)
```
The `data_transform `will eventually contain code for representatives and veteran service organizations as well.  I think the inclusion of that additional code will make this prospective method exceed the length limit.

opticbob commented on modules/representation_management/app/sidekiq/representation_management/accredited_entities_queue_updates.rb:
```diff
@@ -0,0 +1,370 @@
+    # @return [void]
+    def handle_entity_record(entity, config)
+      singular_type = config[:singular_type]
+      entity_hash = send("data_transform_for_#{singular_type}", entity)
```
See https://github.com/department-of-veterans-affairs/vets-api/pull/22911/files#r2180568577

rmtolmach commented:
This is fine as-is. I didn't realize you'd already run this by support (i'm covering support for Lindsey today!)

opticbob commented on modules/representation_management/app/sidekiq/representation_management/accredited_individuals_update.rb:
```diff
@@ -0,0 +1,285 @@
+      record =
+        AccreditedIndividual.find(rep_data['id'])
+      if record.nil?
+        raise StandardError, 'AccreditedIndividual not found.'
```
Done.

opticbob commented:
Monday morning I'm going to convert the multiple references of 'agents' and 'attorneys' to constants to make future changes easier.  I'll also change `singular_type` to `individual_type` in the processing job.
